Moving requests out of execution_payload into beacon_block.body (#3875)
ethereum/consensus-specs: specs/electra/beacon-chain.md

    target_index: ValidatorIndex
```

#### `ExecutionLayerRequests`

> *Note*: This container contains request from the execution layer that are received in the Execution Payload via
> Engine API. These requests are required for CL state transition (see **BeaconBlockBody**).

Reviewer: ```suggestion
*Note*: This container holds requests from the execution layer that are received in [`ExecutionPayloadV4`](https://github.com/ethereum/execution-apis/blob/main/src/engine/prague.md#executionpayloadv4) via
the Engine API. These requests are required for CL state transition (see `BeaconBlockBody`).
```

    target_index: ValidatorIndex
```

#### `ExecutionLayerRequests`

*Note*: This container holds requests from the execution layer that are received in [
`ExecutionPayloadV4`](https://github.com/ethereum/execution-apis/blob/main/src/engine/prague.md#executionpayloadv4) via
the Engine API. These requests are required for CL state transition (see `BeaconBlockBody`).

```python

> class ExecutionLayerRequests(Container):

Reviewer: Personally, I would prefer the name `ExecutionRequests` because:

* It mirrors `ExecutionPayload`.
* We use the field name `execution_requests` in `BeaconBlockBody`.
* And it's shorter.

Author: I'm happy to rename it to `ExecutionPayload`. I'll just wait until others weight in before renaming it in the PR.

    target_index: ValidatorIndex
```

#### `ValidatorRequests`

```python
class ValidatorRequests(Container):

>     deposit_requests: List[DepositRequest, MAX_DEPOSIT_REQUESTS_PER_PAYLOAD]  # [New in Electra:EIP6110]

Reviewer 2: ```suggestion
    deposits: List[DepositRequest, MAX_DEPOSIT_REQUESTS_PER_PAYLOAD]  # [New in Electra:EIP6110]
```

If we have `Requests` suffix in the struct, can we drop it from the fields definition?

    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
    sync_aggregate: SyncAggregate
    # Execution
    execution_payload: ExecutionPayload
    bls_to_execution_changes: List[SignedBLSToExecutionChange, MAX_BLS_TO_EXECUTION_CHANGES]
    blob_kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]

>     requests: ExecutionLayerRequests  # [New in Electra]

Reviewer 2: A thought on making the field name more specific:
```suggestion
    execution_requests: ExecutionLayerRequests  # [New in Electra]
```

Author: Ok, will update :)

    signature: BLSSignature
```

> #### `ExecutionPayload`

Reviewer 2: I believe the `ExecutionPayload` remains unmodified after this PR is applied, so it can be removed from the Electra spec

    withdrawals_root: Root
    blob_gas_used: uint64
    excess_blob_gas: uint64

>     requests_root: Root  # [New in Electra:EIP6110:EIP7002:EIP7251]

Reviewer 2: I believe we don’t need this field at all, as validator requests are already linked to the `BeaconBlockHeader`

            validator.effective_balance = min(balance - balance % EFFECTIVE_BALANCE_INCREMENT, EFFECTIVE_BALANCE_LIMIT)
```

### Execution engine

#### Request data

##### Modified `NewPayloadRequest`

```python
@dataclass
class NewPayloadRequest(object):
    execution_payload: ExecutionPayload
    execution_requests: ExecutionLayerRequests
    versioned_hashes: Sequence[VersionedHash]

>     parent_beacon_block_root: Root

Reviewer: Needs a `# [New in Electra]` comment. And though it makes sense in that location, I'm thinking `execution_requests` should be at the bottom. This makes it easier for clients to extend. What do others think?

```python
def process_block(state: BeaconState, block: BeaconBlock) -> None:
    process_block_header(state, block)

>     process_withdrawals(state, block.body.execution_payload)  # [Modified in Electra:EIP7251:EIP????]

Reviewer 2: ```suggestion
    process_withdrawals(state, block.body.execution_payload)  # [Modified in Electra:EIP7251]
```

>     process_execution_payload(state, block.body, EXECUTION_ENGINE)  # [Modified in Electra:EIP6110:EIP????]

Reviewer 2: ```suggestion
    process_execution_payload(state, block.body, EXECUTION_ENGINE)
```

        state.next_withdrawal_validator_index = next_validator_index
```

#### Execution payload

##### Modified `process_execution_payload`

*Note*: The function `process_execution_payload` is modified to use the new `ExecutionPayloadHeader` type.

```python
def process_execution_payload(state: BeaconState, body: BeaconBlockBody, execution_engine: ExecutionEngine) -> None:
    payload = body.execution_payload

    # Verify consistency of the parent hash with respect to the previous execution payload header
    assert payload.parent_hash == state.latest_execution_payload_header.block_hash
    # Verify prev_randao
    assert payload.prev_randao == get_randao_mix(state, get_current_epoch(state))
    # Verify timestamp
    assert payload.timestamp == compute_timestamp_at_slot(state, state.slot)
    # Verify commitments are under limit
    assert len(body.blob_kzg_commitments) <= MAX_BLOBS_PER_BLOCK
    # Verify the execution payload is valid
    versioned_hashes = [kzg_commitment_to_versioned_hash(commitment) for commitment in body.blob_kzg_commitments]
    assert execution_engine.verify_and_notify_new_payload(
        NewPayloadRequest(

>             execution_payload=payload,

Reviewer 3: we would still need to send the `ExecutionLayerRequests` here for verification

otherwise, there is no way to communicate that the requests derived on the EL matched the ones we apply on the CL

Reviewer: Yeah good catch. I think you're right.

Author: Nice catch! :)

    for_ops(body.deposits, process_deposit)  # [Modified in Electra:EIP7251]
    for_ops(body.voluntary_exits, process_voluntary_exit)  # [Modified in Electra:EIP7251]
    for_ops(body.bls_to_execution_changes, process_bls_to_execution_change)
    for_ops(body.requests.deposits, process_deposit_request)  # [New in Electra:EIP6110]
    # [New in Electra:EIP7002:EIP7251]
    for_ops(body.requests.withdrawals, process_withdrawal_request)
    # [New in Electra:EIP7251]

>     for_ops(body.requests.consolidations, process_consolidation_request)

Reviewer: We should be able to put the "new in ..." comments on the same line now.

>     assert execution_engine.verify_and_notify_new_payload(

Reviewer 4: Is the definition of `verify_and_notify_new_payload()` and `notify_new_payload()` intentionally left out for another PR or coordination with execution-api side of things?

Would need to update these two functions if not.

Author: @ensi321 Do you mean doing the following:
- update function `notify_new_payload` to receive `execution_requests` as a parameter
- update function `verify_and_notify_new_payload` to correctly pass `execution_requests` as a parameter parameter to `notify_new_payload`?

Reviewer 2: IMO, this PR should update `notify_new_payload` call, but there is no need to update Engine API as requests are already incorporated into `newPayloadV4` definition

Author: @ensi321 and @mkalinin I have made the changes. Let me know what you think.

Reviewer 4: > IMO, this PR should update `notify_new_payload` call, but there is no need to update Engine API as requests are already incorporated into `newPayloadV4` definition

As far as I can tell from the two proposals in engine API ethereum/execution-apis#577 ethereum/execution-apis#565, `requests` indeed is part of the `ExecutionPayloadV4` definition. 

But this PR is proposing to remove the 3 requests field from `ExecutionPayload` and store them in `BeaconBlockBody` under `ExecutionRequests`.

So I think we need to update engine API spec to either
1) Add an extra param in `engine_newPayloadV4` for `ExecutionRequests`
2) CL populates `ExecutionPayloadV4.requests` with `ExecutionRequests` before calling `engine_newPayloadV4`

Correct me if I am wrong

Reviewer 2: The design goal of `notify_new_payload` is to make the abstraction layer in the CL spec that would represent the engine. CL spec is agnostic to the specific implementation of the engine, in our case Engine API. In other words, Engine API requests aren’t needed to be reflected in the CL spec, and the spec should rather operate with the structures defined in it. So, `ExecutionRequests` sent to `notify_new_payload` as a separate parameter and sit in the `requests` field of the `ExecutionPayloadV4` looks fine

    for_ops(body.deposits, process_deposit)  # [Modified in Electra:EIP7251]
    for_ops(body.voluntary_exits, process_voluntary_exit)  # [Modified in Electra:EIP7251]
    for_ops(body.bls_to_execution_changes, process_bls_to_execution_change)

>     for_ops(body.requests.deposit_requests, process_deposit_request)  # [New in Electra:EIP6110:EIP????]

Reviewer 2: ```suggestion
    for_ops(body.requests.deposit_requests, process_deposit_request)  # [New in Electra:EIP6110]
```

>     # [New in Electra:EIP7002:EIP7251:EIP????]

Reviewer 2: ```suggestion
    # [New in Electra:EIP7002:EIP7251]
```

    for_ops(body.requests.withdrawal_requests, process_withdrawal_request)

>     # [New in Electra:EIP7251:EIP????]

Reviewer 2: ```suggestion
    # [New in Electra:EIP7251]
```

        transactions_root=hash_tree_root(payload.transactions),
        withdrawals_root=hash_tree_root(payload.withdrawals),
        blob_gas_used=payload.blob_gas_used,

>         excess_blob_gas=payload.excess_blob_gas

Reviewer: We should add back the comma. We like to include commas on last items.
```suggestion
        excess_blob_gas=payload.excess_blob_gas,
```

##### Modified `process_execution_payload`

> *Note*: The function `process_execution_payload` is modified to use the new `ExecutionPayloadHeader` and `ExecutionPayloadEnveloped` types.

Reviewer 2: ```suggestion
*Note*: The function `process_execution_payload` is modified to use the new `ExecutionPayloadHeader` and `ExecutionPayloadEnvelope` types.
```
